[expo-router] - add testing docs (#24045)
expo/expo: docs/pages/router/reference/testing.mdx

---
title: Testing
description: Learn how to use create integration tests for your app.
---

> 

Reviewer: ```suggestion

import { APIBox } from '~/components/plugins/APIBox';
```

Adding APIBox for Jest matchers section.

## Configuration

Before you proceed, ensure you've set up `jest-expo` according to the [Unit Testing documentation](/develop/testing) and [`@testing-library/react-native`](https://callstack.github.io/react-native-testing-library/docs/getting-started).

## renderRouter

`renderRouter` extends the functionality of [`render`](https://callstack.github.io/react-native-testing-library/docs/api#render) to simplify testing with Expo Router. It returns the the same query object as [`render`](https://callstack.github.io/react-native-testing-library/docs/api#render), and is compatible with [`screen`](https://callstack.github.io/react-native-testing-library/docs/api#screen), allowing you to use the standard [query API](https://callstack.github.io/react-native-testing-library/docs/api-queries) to locate components.

`renderRouter` accepts the same [options](https://callstack.github.io/react-native-testing-library/docs/api#render-options) as `render` and introduces an additional option `initialRoute`, which sets an initial route for simulating deep-linking.

### Inline file-system mock

`renderRouter(mock: Record<string, ReactComponent>, options: RenderOptions)`

> `renderRouter` can provide inline-mocking of a file system by passing an object to this function as the first parameter. The keys of the object are the mock filesystem paths. Importantly, when defining these paths, do not use leading relative (`./`) or absolute (`/`) notation and exclude file extension.

Reviewer: ```suggestion
`renderRouter` can provide inline-mocking of a file system by passing an object to this function as the first parameter. The keys of the object are the mock filesystem paths. **Do not use leading relative (`./`) or absolute (`/`) notation when defining these paths and exclude file extension.**
```

We can highlight the context here if it is important for the developer to pay attention.

```tsx app.test.tsx
import { renderRouter, screen } from 'expo-router/testing-library';

it('my-test', async () => {
  const MockComponent = jest.fn(() => <View />);

  renderRouter(
    {
      index: MockComponent,
      'folder/a': MockComponent,
      '(group)/b': MockComponent,
    },
    {
      initialRoute: '/folder/a',
    }
  );

  expect(screen).toHavePathname('/folder/a');
});
```

### Path to a directory

`renderRouter(fixturePath: string, options: RenderOptions)`

`renderRouter` can accept a directory path to mock an existing fixture. Ensure that the provided path is relative to the current test file.

```tsx app.test.js
it('my-test', async () => {
  const MockComponent = jest.fn(() => <View />);
  renderRouter('./my-test-fixture');
});
```

### Path to a directory, with overrides

`renderRouter({ appDir: string, overrides: Record<string, ReactComponent>}, options: RenderOptions)`

> For more intricate testing scenarios, `renderRouter` can leverage both directory path and inline-mocking methods simultaneously. The `appDir` parameter takes a string representing a pathname to a folder. The overrides parameter is an inline mock that can be used to override specific paths within the appDir. This combination allows for fine-tuned control over the mock environment.

Reviewer: ```suggestion
For more intricate testing scenarios, `renderRouter` can leverage both directory path and inline-mocking methods simultaneously. The `appDir` parameter takes a string representing a pathname to a folder. The overrides parameter is an inline mock that can be used to override specific paths within the `appDir`. This combination allows for fine-tuned control over the mock environment.
```

## Jest matchers

> The following matches have been added to `expect` and can be used to asset values on `screen`

Reviewer: ```suggestion
The following matches have been added to `expect` and can be used to asset values on `screen`.
```

> ### `toHavePathname()`

Reviewer: ```suggestion
<APIBox header="toHavePathname()">
```
Let's use APIBox here for better readability. 

![CleanShot 2023-08-24 at 23 39 35](https://github.com/expo/expo/assets/10234615/c52b3d9b-3d55-4d33-87a3-8cee24f78327)

> Assert the current pathname against a given string. The matcher uses the value of the [`usePathname` hook](/router/reference/hooks/#usepathname) on the current `screen`

Reviewer: ```suggestion
Assert the current pathname against a given string. The matcher uses the value of the [`usePathname`](/router/reference/hooks/#usepathname) hook on the current `screen`.
```

```tsx app.test.ts
expect(screen).toHavePathname('/my-router');
```

> 

Reviewer: ```suggestion

</APIBox>
```

> ### `toHaveSegments`

Reviewer: ```suggestion
<APIBox header="toHaveSegments()">
```

> Assert the current segments against an array of strings. The matcher uses the value of the [`useSegments` hook](/router/reference/hooks/#usesegments) on the current `screen`

Reviewer: ```suggestion
Assert the current segments against an array of strings. The matcher uses the value of the [`useSegments`](/router/reference/hooks/#usesegments) hook on the current `screen`.
```

```tsx app.test.ts
expect(screen).toHaveSegments(['[id]']);
```

> 

Reviewer: ```suggestion

</APIBox>
```

> ### `useLocalSearchParams`

Reviewer: ```suggestion
<APIBox header="useLocalSearchParams">
```

Assert the current local search parameters against an object. The matcher uses the value of the [`useLocalSearchParams` hook](/router/reference/hooks/#uselocalsearchparams) on the current `screen`

```tsx app.test.ts
expect(screen).useLocalSearchParams({ first: 'abc' });
```

> 

Reviewer: ```suggestion

</APIBox>
```

> ### `useGlobalSearchParams`

Reviewer: ```suggestion
<APIBox header="useGlobalSearchParams">
```

> Assert the current screens pathname matches a value. Compares using the value of [`useGlobalsearchparams` hook](/router/reference/hooks/#useglobalsearchparams).

Reviewer: ```suggestion
Assert the current screen's pathname that matches a value. Compares using the value of [`useGlobalSearchParams`](/router/reference/hooks/#useglobalsearchparams) hook.
```

> Assert the current global search parameters against an object. The matcher uses the value of the [`useGlobalSearchParams` hook](/router/reference/hooks/#useglobalsearchparams) on the current `screen`

Reviewer: ```suggestion
Assert the current global search parameters against an object. The matcher uses the value of the [`useGlobalSearchParams`](/router/reference/hooks/#useglobalsearchparams) hook on the current `screen`.
```

Does this function have two different functionalities?
